File: app/modules/chrome_debug/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class ChromeDebugTarget(DashboardModel):
    id: str
    type: str | None = None
    title: str | None = None
    url: str | None = None
    attached: bool = False
    browser_context_id: str | None = None
    raw: dict[str, Any] = Field(default_factory=dict)


class ChromeDebugBrowserResponse(DashboardModel):
    id: str
    api_key_id: str
    api_key_name: str | None = None
    label: str
    status: str
    target_count: int = 0
    targets: list[ChromeDebugTarget] = Field(default_factory=list)
    instance_id: str | None = None
    user_agent: str | None = None
    extension_version: str | None = None
    is_revoked: bool = False
    created_at: datetime
    updated_at: datetime
    last_seen_at: datetime | None = None
    disconnected_at: datetime | None = None


class ChromeDebugBrowsersResponse(DashboardModel):
    browsers: list[ChromeDebugBrowserResponse] = Field(default_factory=list)


class ChromeDebugGrantResponse(DashboardModel):
    api_key_id: str
    api_key_name: str
    key_prefix: str
    enabled: bool = False
    browser_count: int = 0
    online_browser_count: int = 0


class ChromeDebugGrantsResponse(DashboardModel):
    grants: list[ChromeDebugGrantResponse] = Field(default_factory=list)


class ChromeDebugGrantUpdateRequest(DashboardModel):
    enabled: bool


class ChromeDebugAgentTokenRequest(DashboardModel):
    browser_id: str | None = Field(default=None, min_length=1, max_length=128)
    label: str = Field(min_length=1, max_length=128)
    user_agent: str | None = None
    extension_version: str | None = Field(default=None, max_length=64)


class ChromeDebugAgentTokenResponse(DashboardModel):
    browser_id: str
    token: str
    expires_at: datetime
    websocket_url: str


class ChromeDebugRelayTokenRequest(DashboardModel):
    browser_id: str = Field(min_length=1)
    ttl_seconds: int = Field(default=300, ge=30, le=3600)


class ChromeDebugDashboardRelayTokenRequest(DashboardModel):
    ttl_seconds: int = Field(default=300, ge=30, le=3600)


class ChromeDebugRelayTokenResponse(DashboardModel):
    token: str
    browser_id: str
    expires_at: datetime
    relay_base_url: str
    json_version_url: str
    json_list_url: str

File: frontend/src/features/chrome-debug/schemas.ts
import { z } from "zod";

export const ChromeDebugTargetSchema = z.object({
  id: z.string(),
  type: z.string().nullable().default(null),
  title: z.string().nullable().default(null),
  url: z.string().nullable().default(null),
  attached: z.boolean().default(false),
  browserContextId: z.string().nullable().default(null),
  raw: z.record(z.string(), z.unknown()).default({}),
});

export const ChromeDebugBrowserSchema = z.object({
  id: z.string(),
  apiKeyId: z.string(),
  apiKeyName: z.string().nullable().default(null),
  label: z.string(),
  status: z.string(),
  targetCount: z.number().int().nonnegative().default(0),
  targets: z.array(ChromeDebugTargetSchema).default([]),
  instanceId: z.string().nullable().default(null),
  userAgent: z.string().nullable().default(null),
  extensionVersion: z.string().nullable().default(null),
  isRevoked: z.boolean().default(false),
  createdAt: z.iso.datetime({ offset: true }),
  updatedAt: z.iso.datetime({ offset: true }),
  lastSeenAt: z.iso.datetime({ offset: true }).nullable().default(null),
  disconnectedAt: z.iso.datetime({ offset: true }).nullable().default(null),
});

export const ChromeDebugBrowsersResponseSchema = z.object({
  browsers: z.array(ChromeDebugBrowserSchema).default([]),
});

export const ChromeDebugGrantSchema = z.object({
  apiKeyId: z.string(),
  apiKeyName: z.string(),
  keyPrefix: z.string(),
  enabled: z.boolean().default(false),
  browserCount: z.number().int().nonnegative().default(0),
  onlineBrowserCount: z.number().int().nonnegative().default(0),
});

export const ChromeDebugGrantsResponseSchema = z.object({
  grants: z.array(ChromeDebugGrantSchema).default([]),
});

export const ChromeDebugRelayTokenResponseSchema = z.object({
  token: z.string(),
  browserId: z.string(),
  expiresAt: z.iso.datetime({ offset: true }),
  relayBaseUrl: z.string(),
  jsonVersionUrl: z.string(),
  jsonListUrl: z.string(),
});

export type ChromeDebugBrowser = z.infer<typeof ChromeDebugBrowserSchema>;
export type ChromeDebugGrant = z.infer<typeof ChromeDebugGrantSchema>;
export type ChromeDebugRelayToken = z.infer<typeof ChromeDebugRelayTokenResponseSchema>;

